crates/types/src/traits/block_contents.rs

```diff
@@ -49,21 +59,33 @@ pub trait BlockPayload:
     type Metadata: Clone + Debug + Eq + Hash + Send + Sync;
 
     /// Encoded payload.
-    type Encode<'a>: 'a + Iterator<Item = u8> + AsRef<[u8]> + Send;
+    type Encode<'a>: 'a + Iterator<Item = u8> + Send
+    where
+        Self: 'a;
 
     /// Build a payload and associated metadata with the transactions.
+    ///
+    /// # Errors
+    /// If the transaction length conversion fails.
     fn from_transactions(
         transactions: impl IntoIterator<Item = Self::Transaction>,
-    ) -> (Self, Self::Metadata);
+    ) -> Result<(Self, Self::Metadata), BlockError>;
```

Comment:
Downstream crates (ie the sequencer) might want more control over what kinds of errors can happen here then just the error types that are predefined in `BlockError`. There are many ways to do this because Rust error handling sucks. I don't have a strong opinion over which is the least bad:
* Make the error an associated type `type Error: Debug + std::error::Error`, then return `Self::Error`
* Make the error something generalized like `anyhow::Error`
* Add a `Custom { message: String }` variant to `BlockError`

Reply:
Ah, actually the first solution was the plan. I've added an associated type `Error: Error + Debug + Send + Sync` to the `Payload` trait, but forgot its existence 😅  and used `BlockError` instead. Will update this!